python_modules/dagster/dagster/_core/execution/context/compute.py

```diff
@@ -1469,27 +1471,65 @@ def job_def(self) -> JobDefinition:
+        For partitioned assets, fetching the latest AssetMaterialization for the partition requires a
+        call to the event log. Additionally, if the currently materializing asset depends on multiple
+        partitions of the upstream asset, ``partition_key`` must be provided so that the latest AssetMaterialization
+        is fetched for the correct partition.
+
         Returns: Optional[AssetMaterialization]
         """
+        asset_key = AssetKey.from_coercible(key)
+
+        if (
+            self.job_def.asset_layer.input_for_asset_key(
+                self.op_execution_context.node_handle, AssetKey.from_coercible(asset_key)
+            )
+            is None
+        ):
+            raise DagsterInvariantViolationError(
+                f"Cannot fetch AssetMaterialization for asset {asset_key}. {asset_key} must be a direct upstream dependency"
+                "in order to call fetch_latest_materialization_for_upstream_asset."
+            )
+
         materialization_events = (
             self.op_execution_context._step_execution_context.upstream_asset_materialization_events  # noqa: SLF001
         )
-        if AssetKey.from_coercible(key) in materialization_events.keys():
-            return materialization_events.get(AssetKey.from_coercible(key))
+        if not self._step_execution_context.has_asset_partitions_for_upstream_asset(asset_key):
+            return materialization_events.get(asset_key)
+        else:
+            # if the asset is partitioned, the AssetMaterialization in 'materialization_events' may not correspond to the
+            # latest AssetMaterialization of the particular partition that is relevant to this materialization. It will
+            # always be the latest AssetMaterialization regardless of partition. So we need to fetch the latest
+            # AssetMaterialization for the specific partition key that is relevant to this materialization
+            if partition_key is None:
+                partition_keys_for_asset = list(
+                    self._step_execution_context.asset_partitions_subset_for_upstream_asset(
+                        asset_key
+                    ).get_partition_keys()
+                )
+                if len(partition_keys_for_asset) > 1:
+                    raise DagsterInvariantViolationError(
+                        f"Asset {self.asset_key} depends on multiple partitions of asset {asset_key}."
+                        " fetch_latest_materialization_for_upstream asset must be given"
+                        " a specific partition_key to fetch the AssetMaterialization for."
+                    )
+                partition_key = partition_keys_for_asset[0]
 
-        raise DagsterInvariantViolationError(
-            f"Cannot fetch AssetMaterialization for asset {key}. {key} must be an upstream dependency"
-            "in order to call latest_materialization_for_upstream_asset."
-        )
+            event = self.instance.get_latest_data_version_record(
```

Comment:
starting a thread to discuss adding this DB access path

Reply:
my major concern with *not* adding this access is that we could be providing inaccurate information to users who are working with partitioned assets

Reply:
and a concern with adding it is that a user could iterate through a big list of assets and fetch the latest materialization for each and break something